File: 표/4_greedy/조이스틱.py
"""
조이스틱으로 알파벳 이름을 완성하세요. 맨 처음엔 A로만 이루어져 있습니다.
- ex) 완성해야 하는 이름이 세 글자면 AAA, 네 글자면 AAAA

▲ - 다음 알파벳
▼ - 이전 알파벳 (A에서 아래쪽으로 이동하면 Z로)
◀ - 커서를 왼쪽으로 이동 (첫 번째 위치에서 왼쪽으로 이동하면 마지막 문자에 커서)
▶ - 커서를 오른쪽으로 이동 (마지막 위치에서 오른쪽으로 이동하면 첫 번째 문자에 커서)
"""

def solution(name):
    answer = 0
    
    min_move = len(name)
    for i, n in enumerate(name):
        # 세로 방향 조작
        if n == 'A':
            pass
        else:
            answer += min(ord('Z') - ord(n) + 1, ord(n) - ord('A'))
        
        # 가로 방향 조작
        move = i + 1
        while move < len(name) and name[move] == 'A':
            move += 1
        
        # 유턴 고려 - ???
        min_move = min(
            min_move,
            i*2 + len(name) - move,
            (len(name) - move)*2 + i
        )

    answer += min_move
        
    return answer

File: 정선웅/조이스틱.java
import java.util.*;

class Solution {
    public int shiftUpDown(char c) {
        int num = (int) c;
        if(num <= 77) {
            return Math.abs(num - 65);
        } else {
            return Math.abs(91 - num);
        }
    }

    public int solution(String name) {
        int answer = 0;
        int length = name.length();

        // 기본 최대 이동 횟수는 그냥 오른쪽으로 끝까지 쭉 가는 경우
        int minMove = length - 1;

        for (int i = 0; i < length; i++) {
            // 1. 알파벳 변경 횟수 누적
            answer += shiftUpDown(name.charAt(i));

            // 2. 현재 위치 i 다음부터 연속된 'A'의 끝자리를 찾음
            int nextIdx = i + 1;
            while (nextIdx < length && name.charAt(nextIdx) == 'A') {
                nextIdx++;
            }

            // 3. 좌우 이동 최소 거리 계산 (3가지 경로 중 최솟값 비교)
            // i * 2 + (length - nextIdx) : 오른쪽 갔다가 빽해서 왼쪽으로
            // (length - nextIdx) * 2 + i : 왼쪽 먼저 갔다가 빽해서 오른쪽으로
            minMove = Math.min(minMove, i * 2 + (length - nextIdx));
            minMove = Math.min(minMove, (length - nextIdx) * 2 + i);
        }

        return answer + minMove;
    }
}
